File: src/iOS/CustomPresetController.m
//
//  CustomPresetController.h
//  Go Map!!
//
//  Created by Bryce Cogswell on 8/21/14.
//  Copyright (c) 2014 Bryce Cogswell. All rights reserved.
//

#import "CommonPresetList.h"
#import "CustomPresetController.h"
#import "POIFeaturePresetsViewController.h"

@interface CustomPresetController ()
@end

@implementation CustomPresetController

- (void)viewDidLoad
{
	[super viewDidLoad];

	_valueFieldList = @[ _value1Field, _value2Field, _value3Field, _value4Field,  _value5Field,  _value6Field,
						 _value7Field, _value8Field, _value9Field, _value10Field, _value11Field, _value12Field ];

	_nameField.text				= _customPreset.name;
	_appliesToTagField.text		= _customPreset.appliesToKey;
	_appliesToValueField.text	= _customPreset.appliesToValue;
	_keyField.text				= _customPreset.tagKey;

	NSInteger idx = 0;
	for ( UITextField * textField in _valueFieldList ) {
		if ( idx >= _customPreset.presetList.count )
			break;
		CommonPresetValue * preset = _customPreset.presetList[ idx ];
		textField.text = preset.tagValue;
		++idx;
	}
}


-(IBAction)done:(id)sender
{
	// remove white space from subdomain list
	NSString * name = [_nameField.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
	NSString * key  = [_keyField.text  stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
	if ( name.length == 0 || key.length == 0 )
		return;
	NSString * appliesToKey = [_appliesToTagField.text	 stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
	NSString * appliesToVal = [_appliesToValueField.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
	NSMutableArray * presets = [NSMutableArray new];
	for ( UITextField * field in _valueFieldList ) {
		NSString * value = [field.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
		if ( value.length ) {
			CommonPresetValue * preset = [CommonPresetValue presetWithName:nil details:nil tagValue:value];
			[presets addObject:preset];
		}
	}
	UIKeyboardType keyboard = UIKeyboardTypeDefault;
	UITextAutocapitalizationType capitalize = UITextAutocapitalizationTypeNone;
	
	_customPreset = [[CustomPreset alloc] initWithName:name tagKey:key defaultValue:nil placeholder:nil keyboard:keyboard capitalize:capitalize presets:presets];
	_customPreset.appliesToKey = appliesToKey;
	_customPreset.appliesToValue = appliesToVal;
	if ( _completion ) {
		_completion(_customPreset);
	}
	[self.navigationController popViewControllerAnimated:YES];
}

-(IBAction)cancel:(id)sender
{
	[self.navigationController popViewControllerAnimated:YES];
}

-(IBAction)contentChanged:(id)sender
{
	if ( _nameField.text.length > 0 && _keyField.text.length > 0 ) {
		self.navigationItem.rightBarButtonItem.enabled = YES;
	} else {
		self.navigationItem.rightBarButtonItem.enabled = NO;
	}
}


@end

File: src/iOS/CustomPresetController.swift
//
//  CustomPresetController.swift
//  Go Map!!
//
//  Created by Bryce Cogswell on 8/21/14.
//  Copyright (c) 2014 Bryce Cogswell. All rights reserved.
//

import UIKit

class CustomPresetController: UITableViewController {
	@IBOutlet var nameField: UITextField!
	@IBOutlet var appliesToTagField: UITextField!
	@IBOutlet var appliesToValueField: UITextField!
	@IBOutlet var keyField: UITextField!
	@IBOutlet var value1Field: UITextField!
	@IBOutlet var value2Field: UITextField!
	@IBOutlet var value3Field: UITextField!
	@IBOutlet var value4Field: UITextField!
	@IBOutlet var value5Field: UITextField!
	@IBOutlet var value6Field: UITextField!
	@IBOutlet var value7Field: UITextField!
	@IBOutlet var value8Field: UITextField!
	@IBOutlet var value9Field: UITextField!
	@IBOutlet var value10Field: UITextField!
	@IBOutlet var value11Field: UITextField!
	@IBOutlet var value12Field: UITextField!
	private var valueFieldList: [UITextField] = []

	var customPreset: PresetKeyUserDefined?
	var completion: ((_ customPreset: PresetKeyUserDefined?) -> Void)?

	@IBAction func contentChanged(_ sender: Any) {
		if (nameField.text?.count ?? 0) > 0, (keyField.text?.count ?? 0) > 0 {
			navigationItem.rightBarButtonItem?.isEnabled = true
		} else {
			navigationItem.rightBarButtonItem?.isEnabled = false
		}
	}

	@IBAction func done(_ sender: Any) {
		// remove white space from subdomain list
		let name = nameField.text?.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines) ?? ""
		let key = keyField.text?.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines) ?? ""
		if name.count == 0 || key.count == 0 {
			return
		}
		let appliesToKey = appliesToTagField.text?.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines) ?? ""
		let appliesToVal = appliesToValueField.text?.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines) ?? ""
		var presets: [PresetValue] = []
		for field in valueFieldList {
			let value = field.text?.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines) ?? ""
			if value.count != 0 {
				let preset = PresetValue(name: nil, details: nil, tagValue: value)
				presets.append(preset)
			}
		}
		let keyboard: UIKeyboardType = .default
		let capitalize: UITextAutocapitalizationType = .none

		customPreset = PresetKeyUserDefined(
			appliesToKey: appliesToKey,
			appliesToValue: appliesToVal,
			name: name,
			tagKey: key,
			placeholder: nil,
			keyboard: keyboard,
			capitalize: capitalize,
			presets: presets)
		completion?(customPreset)
		navigationController?.popViewController(animated: true)
	}

	@IBAction func cancel(_ sender: Any) {
		navigationController?.popViewController(animated: true)
	}

	override func viewDidLoad() {
		super.viewDidLoad()

		valueFieldList = [
			value1Field,
			value2Field,
			value3Field,
			value4Field,
			value5Field,
			value6Field,
			value7Field,
			value8Field,
			value9Field,
			value10Field,
			value11Field,
			value12Field
		]

		nameField.text = customPreset?.name ?? ""
		appliesToTagField.text = customPreset?.appliesToKey ?? ""
		appliesToValueField.text = customPreset?.appliesToValue ?? ""
		keyField.text = customPreset?.tagKey ?? ""

		var idx = 0
		for textField in valueFieldList {
			if idx >= (customPreset?.presetList?.count ?? 0) {
				break
			}
			let preset = customPreset?.presetList?[idx]
			textField.text = preset?.tagValue
			idx += 1
		}
	}
}
